src/mono/wasm/runtime/strings.ts

```diff
@@ -272,3 +269,68 @@ export function js_string_to_mono_string_new(string: string): MonoString {
         temp.release();
     }
 }
+
+let _text_decoder_utf8_relaxed: TextDecoder | undefined = undefined;
+let _text_decoder_utf8_validating: TextDecoder | undefined = undefined;
+let _text_encoder_utf8: TextEncoder | undefined = undefined;
+
+export function encodeUTF8(str: string): Uint8Array {
+    if (_text_encoder_utf8 === undefined) {
+        if (typeof TextEncoder === "undefined") {
+            const buffer = new Uint8Array(str.length * 2);
+            Module.stringToUTF8Array(str, buffer, 0, str.length * 2);
+            return buffer;
+        }
+        _text_encoder_utf8 = new TextEncoder();
+    }
+    return _text_encoder_utf8.encode(str);
+}
+
+export function utf8ToStringRelaxed(buffer: Uint8Array): string {
+    if (_text_decoder_utf8_relaxed === undefined) {
+        if (typeof TextDecoder === "undefined") {
+            return Module.UTF8ArrayToString(buffer, 0, buffer.byteLength);
+        }
+        _text_decoder_utf8_relaxed = new TextDecoder("utf-8", { fatal: false });
+    }
+    return _text_decoder_utf8_relaxed.decode(buffer);
+}
+
+export function utf8ToString(ptr: CharPtr): string {
+    updateGrowableHeapViews();
+    return decodeUTF8(Module.HEAPU8, ptr as any, Module.HEAPU8.length - (ptr as any));
+}
+
+export function decodeUTF8(heapOrArray: Uint8Array, idx: number, maxBytesToRead: number): string {
+    const endIdx = idx + maxBytesToRead;
+    let endPtr = idx;
+    while (heapOrArray[endPtr] && !(endPtr >= endIdx)) ++endPtr;
+    if (endPtr - idx <= 16) {
+        return Module.UTF8ArrayToString(heapOrArray, idx, maxBytesToRead);
+    }
+    if (_text_decoder_utf8_validating === undefined) {
+        if (typeof TextDecoder === "undefined") {
+            return Module.UTF8ArrayToString(heapOrArray, idx, maxBytesToRead);
+        }
+        _text_decoder_utf8_validating = new TextDecoder("utf-8");
+    }
+    const view = copy_string_buffer_as_necessary(heapOrArray, idx as any, endPtr as any);
+    return _text_decoder_utf8_validating.decode(view);
+}
+
+const mayNeedCopy = MonoWasmThreads && typeof SharedArrayBuffer !== "undefined";
+
+// When threading is enabled, TextDecoder does not accept a view of a
+// SharedArrayBuffer, we must make a copy of the array first.
+// See https://github.com/whatwg/encoding/issues/172
+// BEWARE: In some cases, `instanceof SharedArrayBuffer` returns false even though buffer is an SAB.
+// Patch adapted from https://github.com/emscripten-core/emscripten/pull/16994
+// See also https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag
+export function copy_string_buffer_as_necessary(view: Uint8Array, start: CharPtr, end: CharPtr): Uint8Array {
+    updateGrowableHeapViews();
+    // this condition should be eliminated by rollup on non-threading builds
+    const needsCopy = mayNeedCopy && view.buffer[Symbol.toStringTag] === "SharedArrayBuffer";
```

Comment:
Let's make a general purpose utility

```ts
function isSharedArrayBuffer(x: any): x is SharedArrayBuffer
{
  return mayNeedCopy && view.buffer[Symbol.toStringTag] === "SharedArrayBuffer"
}

Reply:
This is what `StringDecoder.decode` boiled down to in non-MT build

![image](https://github.com/dotnet/runtime/assets/271576/cc2fd1b4-d829-4ca6-ac44-e25e7e01be9d)